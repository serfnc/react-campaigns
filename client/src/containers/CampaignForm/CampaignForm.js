import React from "react";

import {Route, BrowserRouter as Router, Switch, NavLink} from "react-router-dom";

import request from "superagent";

import moment from 'moment';
import Cleave from 'cleave.js/react';


class CampaignForm extends React.Component {
    _isMounted = false;

    constructor(props) {
        super(props);

        this.state = {
            loading: false,
            id: props.match && props.match.params && props.match.params.id ? props.match.params.id : null,
            campaign: props.location && props.location.campaign ? props.location.campaign : {
                id: '',
                startDate: '',
                endDate: '',
                targetImpressions: ''
            }
        };
    }

    componentDidMount() {
        this._isMounted = true;

        if (this.state.id && !this.state.campaign.id) {
            this.setState({loading: true}, () => {
                this.getCampaign()
            })
        } else if (this.state.campaign) {
            this.repairCampaign()
        }
    }

    componentWillUnmount() {
        this._isMounted = false;
    }


    repairCampaign() {
        var currentCampaign = this.state.campaign

        if (this.state.campaign.startDate) {
            currentCampaign.startDate = moment(currentCampaign.startDate).format('DD/MM/YYYY')
            this.setState({campaign: currentCampaign})
        }

        if (this.state.campaign.endDate) {
            currentCampaign.endDate = moment(currentCampaign.endDate).format('DD/MM/YYYY')
            this.setState({campaign: currentCampaign})
        }
    }

    getCampaign() {
        this.setState({loading: true}, () => {
            request
                .get('http://localhost:8000/api/get-campaign/' + this.state.id)
                .then((results) => {
                    this.setState({
                        loading: false,
                        campaign: results.body.campaign
                    })
                    this.repairCampaign()
                })
                .catch((err) => {
                    //console.log(err)

                    this.setState({
                        loading: false,
                    });
                })
        })
    }

    handleSubmit(e) {
        e.preventDefault()

        request
            .post('http://localhost:8000/api/save-campaign')
            .send({campaign: this.state.campaign})
            .set('Content-Type', 'application/x-www-form-urlencoded')
            .set('Accept', 'application/json')
            .then((result) => {
                console.log(result)
                this.props.history.push('/')
            })
            .catch((err) => {
                if (err.response && err.response.body) {
                    alert(err.response.body)
                } else {
                    alert(err)
                }
            })
    }

    handleCancel(e) {
        this.props.history.push('/')
    }

    handleStartDateChange(e) {
        var newCampaign = this.state.campaign
        newCampaign.startDate = e.target.value
        this.setState({campaign: newCampaign})
    }

    handleEndDateChange(e) {
        var newCampaign = this.state.campaign
        newCampaign.endDate = e.target.value
        this.setState({campaign: newCampaign})
    }

    handleTargetChange(e) {
        var newCampaign = this.state.campaign
        newCampaign.targetImpressions = e.target.value
        this.setState({campaign: newCampaign})
    }

    render() {
        if (this.state.loading) {
            return (
                <div className="container is-text-center" style={{padding: "100px 20px"}}>
                    <p>Loading...</p>
                </div>
            )
        }

        return (
            <div className="container" style={{padding: "100px 20px"}}>
                <div className="is-50 is-center">
                    <div className="page-header">
                        <h1>{this.state.id ? 'Edit campaign' : 'New campaign'}</h1>
                        <p>Add the campaign details.</p>
                    </div>

                    <div className="page-content">
                        <form method="post" onSubmit={this.handleSubmit.bind(this)} action="">
                            <div className="form-item">
                                <label>Start date</label>
                                <Cleave options={{date: true, delimiter: '/', datePattern: ['d', 'm', 'Y']}} placeholder="dd/mm/yyyy" type="text" name="startDate" value={this.state.campaign.startDate} onChange={this.handleStartDateChange.bind(this)} className=""/>
                            </div>

                            <div className="form-item">
                                <label>End date</label>
                                <Cleave options={{date: true, delimiter: '/', datePattern: ['d', 'm', 'Y']}} placeholder="dd/mm/yyyy" type="text" name="endDate" value={this.state.campaign.endDate} onChange={this.handleEndDateChange.bind(this)} className=""/>
                            </div>

                            <div className="form-item">
                                <label>Target impressions</label>
                                <Cleave options={{numericOnly: true}} type="text" name="targetImpressions" value={this.state.campaign.targetImpressions} placeholder="1000" onChange={this.handleTargetChange.bind(this)} className=""/>
                            </div>

                            <div className="form-item is-buttons">
                                {!this.state.id &&
                                    <button className="button">{this.state.campaign.id ? 'Update campaign' : 'Add campaign'}</button>
                                }
                                <button onClick={this.handleCancel.bind(this)} className="button is-tertiary">Cancel</button>
                            </div>

                        </form>
                    </div>
                </div>
            </div>
        );
    }
}

export default CampaignForm;
