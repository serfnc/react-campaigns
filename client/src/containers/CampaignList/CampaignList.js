import React from "react";

import {Route, BrowserRouter as Router, Switch, NavLink} from "react-router-dom";

import Campaign from "./../../components/Campaign/Campaign";

import request from "superagent";

class CampaignList extends React.Component {
    _isMounted = false;

    constructor(props) {
        super(props);

        this.state = {
            loading: false,
            campaigns: []
        };
    }

    componentDidMount() {
        this._isMounted = true;

        this.getCampaigns()
    }

    componentWillUnmount() {
        this._isMounted = false;
    }


    getCampaigns() {
        this.setState({loading: true}, () => {
            request
                .get('http://localhost:8000/api/get-campaigns')
                .then((results) => {
                    this.setState({
                        loading: false,
                        campaigns: results.body.campaigns ? results.body.campaigns : []
                    })
                })
                .catch((err) => {
                    this.setState({
                        loading: false,
                    });
                })
        })
    }

    render() {
        var pageContent = ''

        if (this.state.loading) {
            pageContent = (
                <div className="container is-text-center" style={{padding: "100px 20px"}}>
                    <p>Loading...</p>
                </div>
            )
        } else {
            pageContent = (
                <div>
                    {this.state.campaigns.map(campaign => (
                        <Campaign key={campaign.id} campaign={campaign}/>
                    ))}
                </div>
            )
        }

        return (
            <div className="container" style={{padding: "100px 20px"}}>
                <div className="page-header is-row">
                    <div className="is-col">
                        <h1>Campaigns</h1>
                        <p>Find below the list of all active campaigns.</p>
                    </div>
                    <div className="is-col is-text-right">
                        <NavLink to={{pathname: "/campaign/new"}} className="button add-campaign">Add campaign</NavLink>
                    </div>
                </div>

                <div className="page-content">
                    {pageContent}
                </div>
            </div>
        );
    }
}

export default CampaignList;