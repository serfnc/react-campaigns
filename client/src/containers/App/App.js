import React from 'react';

import {Route, BrowserRouter as Router, Switch} from "react-router-dom";

import Header from "./../../components/Header/Header";

import CampaignList from "./../CampaignList/CampaignList";
import CampaignForm from "./../CampaignForm/CampaignForm";

import style from './../../sass/kube.scss';

function App() {
    return (
        <div>
            <Header/>
            <Router>
                <Switch>
                    <Route exact path="/" component={CampaignList}/>
                    <Route exact path="/campaign/new" component={CampaignForm}/>
                    <Route path="/campaign/:id" component={CampaignForm}/>
                </Switch>
            </Router>
        </div>
    );
}

export default App;
