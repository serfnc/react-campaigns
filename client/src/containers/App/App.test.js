import React from 'react';
import ReactDOM from "react-dom";
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import App from './App';
import waitUntil from "async-wait-until";
import CampaignList from "../CampaignList/CampaignList";
import CampaignForm from "../CampaignForm/CampaignForm";

it('matches the snapshot', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders without crashing', () => {
    shallow(<App />);
});
/*

it('check if results render correctly', async (done) => {
    const wrapper = shallow(<App />);

    await waitUntil(() => wrapper.find('div.page-content').text().length > 0);

    console.log(wrapper.html())

    /!*wrapper.find('.campaign').forEach( (node) => {
        console.log(node.html())
        expect(node.find('.campaign-title').text()).toEqual(true)
    })*!/

    done();
});*/
