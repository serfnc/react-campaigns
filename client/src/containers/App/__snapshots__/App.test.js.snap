// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div>
  <header
    className="header"
  >
    <a
      href="/"
    >
      <span>
        Campaigns
      </span>
      <img
        alt="Campaigns"
        src="logo.svg"
      />
    </a>
  </header>
  <div
    className="container"
    style={
      Object {
        "padding": "100px 20px",
      }
    }
  >
    <div
      className="page-header is-row"
    >
      <div
        className="is-col"
      >
        <h1>
          Campaigns
        </h1>
        <p>
          Find below the list of all active campaigns.
        </p>
      </div>
      <div
        className="is-col is-text-right"
      >
        <a
          aria-current={null}
          className="button add-campaign"
          href="/campaign/new"
          onClick={[Function]}
        >
          Add campaign
        </a>
      </div>
    </div>
    <div
      className="page-content"
    >
      <div
        className="container"
        style={
          Object {
            "padding": "100px 20px",
          }
        }
      >
        <p>
          Loading...
        </p>
      </div>
    </div>
  </div>
</div>
`;
